Gsoc2019 kdop tree xiao

xiaoXiaoCam opened the pull request:
K-dop computation by traversing the tree.

afabri committed:
Add KDOP_tree to the manual

Commit:
Typos in header files.

Commit:
include < > header files (tmp).

Commit:
Doxygen package description updated (exercise).

Commit:
Updated API structure for k-dop.

Commit:
Implemented and tested the k-dop computation by the union of the child
nodes. Need to further test and clean up the code.

Commit:
Added the case "nb_primitives = 3" in kdop_traversal.

Commit:
Integrate "compute_support_heights" of class KDOP_kdop into
"compute_kdop" of class KDOP_traits.

afabri commented on KDOP_tree/include/CGAL/KDOP_tree/internal/Has_nested_type_Shared_data.h:
```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2019  University of Cambridge (UK), GeometryFactory (France)
```
There is no need for this file as nothing is kdop specific.  Just include the corresponding file from the AABB_tree package.

xiaoXiaoCam replied:
This file has been removed and replaced by the corresponding AABB_tree file.

afabri commented on KDOP_tree/doc/KDOP_tree/Concepts/KDOPKdop.h:
```diff
@@ -0,0 +1,95 @@
+    /*!
+     * Return support heights
+     */
+    inline Vec_height give_support_heights() const;
```
For data accessors it is a convention to  not prefix with `get_`  or as you do with `give_`.
Also it should return a `const&` so that we do not copy the `std::vector`.
If the `K` was a template parameter and hence constant you could use a `std::array<double,K>` instead of the `std::vector>.

xiaoXiaoCam replied:
Has changed to "Vec_height support_heights() const&".

xiaoXiaoCam replied:
Added a template parameter N to the class, std::vector<double> is replaced by std::array<double, N>.

afabri commented on KDOP_tree/include/CGAL/KDOP_tree/KDOP_kdop.h:
```diff
@@ -0,0 +1,181 @@
+    bool do_overlap(const Kdop& kdop1, const Kdop& kdop2);
+
+  private:
+    Vec_direction vector_direction_;
```
I do not understand why each kdop stores a vector with directions.  For the kdops in one tree they should all be the same.

xiaoXiaoCam replied:
Yes, they should be the same. Maybe I can modify this so that the direction vector is imported as a parameter in member functions that require directions.

xiaoXiaoCam replied:
Adapted the member functions so that it is not necessary to store direction vector in each kdop.

Commit:
Modified the code according to the reviews: add template parameter N for
k-dops; direction vector is not necessary; the support heights are
stored in an array of size N.

Commit:
Do_intersect with a test .cpp file.

afabri commented on KDOP_tree/include/CGAL/KDOP_tree/KDOP_tree.h:
```diff
@@ -241,7 +241,7 @@ namespace KDOP_tree {
     /// defined in the traits class `KDOPTraits`.
     template<typename Query>
-    bool do_intersect(const Query& query) const;
+    bool do_intersect(const Query& query);
```
Why do you have to remove constness?  This is only a test.

xiaoXiaoCam replied:
It didn't compile with 'const' last night... I just read the code again and fixed this problem.

Commit:
Fixed the 'const' issue for Do_intersect.

afabri commented on KDOP_tree/include/CGAL/KDOP_tree/KDOP_kdop.h:
```diff
@@ -151,21 +150,24 @@ namespace KDOP_tree {
 
-    std::array<double, direction_number> support_heights1 = kdop1.support_heights();
-    std::array<double, direction_number> support_heights2 = kdop2.support_heights();
+    std::array<double, direction_number> support_heights = this->support_heights();
```
Should be `const Array_height&`  in order to avoid a copy.

afabri commented on KDOP_tree/include/CGAL/KDOP_tree/KDOP_kdop.h:
```diff
@@ -0,0 +1,185 @@
+      double height2 = support_heights_query[i];
+
+      if (height1 + height2 < 0) {
+        num_non_overlap += 1;
```
You could write a `return` directly here, instead of incrementing, breaking, testing.

Commit:
Ray intersection. The code structure needs improvement.

afabri commented on KDOP_tree/include/CGAL/KDOP_tree/KDOP_kdop.h:
```diff
@@ -0,0 +1,259 @@
+    bool do_overlap_segment(const Query& q, const Vec_direction& directions) const;
+
+    // Ray
+    // \todo support heights of the query should be located before traversal (need discussion)
```
What is the support height of a ray.  Can it be infinity, as it is unbounded?

xiaoXiaoCam replied:
For a ray, the k-dop is unbounded, so we need to use the heights at the two points (a source point and the other point on the ray) and compare them with the support heights of the object.

afabri replied:
And is it correct to just take another point on the ray, that is does it not matter how far the second point is away from the source???

xiaoXiaoCam replied:
Yes, the premise is that the height of the second point is larger than that of the source point in that direction. Thanks to the fact that the directions we consider are always in pairs, this premise can always satisfy in one of the directions in pairs. I can upload an illustration on wiki later.

afabri replied:
What do you mean with "direction in pairs"?  parallel slabs?  If there any premises they must be documented.

xiaoXiaoCam replied:
Like +x and -x, etc. I can document this with more description and explanation.

afabri commented on KDOP_tree/include/CGAL/KDOP_tree/KDOP_kdop.h:
```diff
@@ -0,0 +1,259 @@
+
+
+  private:
+    Vec_direction vector_direction_; // Redundant at the moment
```
You should not wait to remove this data member, and all occurrences where it is used.

afabri commented on KDOP_tree/include/CGAL/KDOP_tree/KDOP_primitive.h:
```diff
@@ -0,0 +1,263 @@
+// Copyright (c) 2019  University of Cambridge (UK), GeometryFactory (France)
```
Is anything different from the same class of the AABB tree package? I don't see a bbox, so I guess there is no need to copy/rename, but plwease check again.

afabri commented on KDOP_tree/include/CGAL/KDOP_tree/internal/KDOP_node.h:
```diff
@@ -0,0 +1,388 @@
+
+    // node kdop
+    Kdop m_kdop;
+    Vec_direction m_directions;
```
The node class also should NOT store the directions.

afabri commented on KDOP_tree/include/CGAL/KDOP_tree/internal/KDOP_node.h:
```diff
@@ -0,0 +1,388 @@
+  KDOP_node<Tr>::union_support_heights(const Array_height& left_heights,
+                                       const Array_height& right_heights,
+                                       Array_height& heights_union,
+                                       const int direction_number)
```
No need for `direction_number`.  Should be an `unsigned int` template parameter.  Maybe go through the entire code to clean that up in a single commit.

Commit:
Test file to compare do_intersect of rays using AABB and KDOP.

Commit:
Cleaned up the code to remove unnecessary 'num_directions' arguments and
redundant 'm_directions' variables.

Commit:
Added a function to read the point file .xyz in test files; created a
new test file to compare the performance of AABB and KDOP.

afabri committed:
Add timing

Commit:
Used 'query_pair' to pair the query and its kdop in order to use
precomputed kdop of the query.

afabri committed:
Use const&

Commit:
Improved ray intersection criterion considering ray/plane intersection.
This is far more efficient.

Commit:
Minor changes.

afabri committed:
Merge branch 'gsoc2019-KDOP_tree-xiao' of github.com:CGAL/cgal-public-dev into gsoc2019-KDOP_tree-xiao

afabri committed:
VC++ wants && and ||;   Some const&

Commit:
Test file for 6-dop.

Commit:
Merge remote-tracking branch 'refs/remotes/origin/gsoc2019-KDOP_tree-xiao' into gsoc2019-KDOP_tree-xiao

Commit:
Minor changes (comment out the counter output as it requires changes in
other files).

Commit:
Visualise the kdop with convex hulls by halfspace intersection (1st
version)

afabri committed:
no std::cout during performance comparison;  CGAL_PI

Commit:
Improved k-dop by combining computation of opposite directions.

Commit:
Plate test (without node counter)

Commit:
Test with computing heights directly without dot product; changed the
checking order in overlap check.

Commit:
Minor change (commented out a line by mistake in the last commit)

Commit:
node.h to store support_heights directly; kdop.h in progress.

Commit:
Support height computation is hardcoded for N = 6, 14, 18 and 26.

Commit:
Generalise computing support heights. \todo generalise overlap checking.

Commit:
Generalised overlap checking to call function based on query type.

Commit:
Distance query in progress. \todo function to compare distances in
traits class.

Commit:
Add function to compare distances; need to test and debug.

Commit:
Debugged distance query; used unit directions to compare distances.

Commit:
Minor changes to closest_point query; add points file for test.

Commit:
Updated test files.

afabri commented on KDOP_tree/include/CGAL/KDOP_tree/KDOP_kdop.h:
```diff
@@ -0,0 +1,449 @@
+      }
+    }
+
+    if (distance > squared_radius) return false;
```
Must this test not also be done after the next for loop?
And why is it not done after the  `distance +=`  with distance IN the two loops?

xiaoXiaoCam replied:
For a k-dop, it is not straightforward to compute the distance between a point and a polytope (considering the polytope is the intersections of k half-planes), so this distance is not used for k>6 and it is simplified to compare only the support heights in each direction. This may leave some regions with false positives, but those regions are very small. The exact computation of these regions may not have benefit in reducing the time.

The 'distance' is the distance between the point and the bbox. We could also put this check 'distance > squared_radius' in the loops right after 'distance += ...'.

Commit:
KDOP computation of primitives in a new class Construct_kdop.

Commit:
Removed some files same as AABB; added more ray query functions.

Commit:
Added first_intersection & more distance query functions.

Commit:
Test file to check query correctness with k-dops.

Commit:
Clean-up of the code and examples.

Commit:
New CMakeLists.txt

Commit:
Manual files 1.

Commit:
Manual editing
